src/main/java/seedu/address/model/loan/DataField.java

```diff
@@ -0,0 +1,72 @@
+package seedu.address.model.loan;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * Represents a generic field that is directly describable as a string (called an objString).
+ * T: The type of the underlying value this DataField has.
+ * Guarantees: immutable; is valid as per the isValid function passed to the constructor.
+ */
+public abstract class DataField<T> {
+
+    /**
+     * Subclasses should have:
+     * - {@code static final String MESSAGE_CONSTRAINTS = [string];}
+     *      A message to the user describing how a valid objString should be formatted.
+     * - {@code static final Predicate<String> VALIDITY_PREDICATE = test -> test.matches([regex-expression]);}
+     *      Returns true iff the given objString is valid.
+     * - The constructor only needs to be of this format:
+     *     - arguments: {@code String objString}
+     *     - {@code { super(MESSAGE_CONSTRAINTS, VALIDITY_PREDICATE, parser, objString); }}
+     *         - The parser converts the given objString into an object of type {@code T}, e.g. {@code Integer.parseInt}
+     *         - The parser must be guaranteed to work (i.e. output a valid value) on a valid objString
+     */
+
+    public final T value;
+
+    /**
+     * Constructs a {@code LoanField}.
+     *
+     * @param msgConstraints A message to the user describing how a valid objString should be formatted.
+     * @param validityPred A predicate to check if the given objString is valid.
```

Comment:
Shouldn't validation be done in the class level instead of having to explicitly pass into the constructor? I am not sure about this.